pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import torch
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import torch_ones_like, torch_zeros_like
+from pyro.infer.mcmc.trace_kernel import TraceKernel
+from pyro.ops.integrator import velocity_verlet
+from pyro.util import ng_ones, ng_zeros
+
+
+class HMC(TraceKernel):
+    """
+    Simple Hamiltonian Monte Carlo kernel, where `step_size` and `num_steps`
+    need to be explicitly specified by the user.
+
+    **Reference**
+    "MCMC Using Hamiltonian Dynamics", R Neal. `pdf <https://arxiv.org/pdf/1206.1901.pdf>`_
+
+
+    :param model: python callable containing pyro primitives.
+    :param float step_size: determines the size of a single step taken by the
+        verlet integrator while computing the trajectory using Hamiltonian
+        dynamics.
+    :param int num_steps: The number of discrete steps over which to simulate
+        Hamiltonian dynamics. The state at the end of the trajectory is
+        returned as the proposal.
+    """
+
+    def __init__(self, model, step_size=0.5, num_steps=3):
+        self.model = model
+        self.step_size = step_size
+        self.num_steps = num_steps
+        # simulation run attributes - will be set in self._setup at start of run
+        self._t = 0
```

Comment:
I'd recommend calling `self.cleanup()` here rather than duplicating it's logic. That eliminates the need to keep these two pieces of code synchronized. Also consider renaming `cleanup` to `reset` so it makes more sense to call here?

Reply:
Makes sense. The only thing is that `setup` and `cleanup` are methods in the abstract `TraceKernel` class. I suppose we could just change the interface to have a single `reset` method. I don't currently foresee us needing different methods at this point.